### PR TITLE
Adds grpc-based thread-local support

### PR DESCRIPTION
**Pull request by dinooliva**

### CODE REVIEW

**Comment by sebright** [comment]
It looks like this branch continues from dinooliva/thread-local, but I think the merge will still work.

File: examples/java/com/google/census/CensusRunner.java [hunk 1]
```diff
@@ -36,11 +39,42 @@
 
   public static void main(String args[]) {
     System.out.println("Hello Census World");
-    System.out.println("Default Tags: " + Census.getCensusContextFactory().getDefault());
```
Comment (adriancole) [inline examples/java/com/google/census/CensusRunner.java]: Nit: hard to see the changes due to extracting methods. Presumably biggest change here is catch blocks

File: grpc/java/com/google/census/CensusGrpcContext.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+package com.google.census;
+
+import io.grpc.Context;
+
+/**
+ * Utils for embedding {@link CensusContext} in gRPC {@link Context} to support thread-local
+ * implicit propagation of the {@link CensusContext}.
+ */
+public abstract class CensusGrpcContext {
```
Comment (adriancole) [inline grpc/java/com/google/census/CensusGrpcContext.java]: Did you want this ctor public? Or is it intended as package private?

File: grpc_native/java/com/google/census/CensusGrpcContextImpl.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+package com.google.census;
+
+import io.grpc.Context;
+
+/**
+ * Utils for embedding {@link CensusContext} in gRPC {@link Context} to support thread-local
+ * implicit propagation of the {@link CensusContext}.
+ */
+public class CensusGrpcContextImpl extends CensusGrpcContext {
```
Comment (adriancole) [inline grpc_native/java/com/google/census/CensusGrpcContextImpl.java]: Intentionally not final?

File: examples/java/com/google/census/CensusRunner.java (hunk 1 above)
Comment (sebright) [inline examples/java/com/google/census/CensusRunner.java]: Do you mean that the PR should have been broken into multiple commits to make the diffs clearer?

The main difference is that the new code uses the gRPC Context to keep track of the Census tags.  We could add back the `CensusScope` class, implemented using the gRPC context, to make this code clearer.

File: grpc/java/com/google/census/CensusGrpcContext.java (hunk 2 above)
Comment (sebright) [inline grpc/java/com/google/census/CensusGrpcContext.java]: This class is supposed to act as an interface that can have methods added in Java 6 and 7, so the constructor is intentionally left public.

File: grpc_native/java/com/google/census/CensusGrpcContextImpl.java (hunk 3 above)
Comment (sebright) [inline grpc_native/java/com/google/census/CensusGrpcContextImpl.java]: No.  I fixed it in #27.

**Pull request merged**